src/sage/schemes/elliptic_curves/ell_finite_field.py

```diff
@@ -1122,14 +1122,53 @@ def torsion_basis(self, n):
         :meth:`AdditiveAbelianGroupWrapper.torsion_subgroup`.
         """
         # TODO: In many cases this is not the fastest algorithm.
-        # Alternatives include factoring division polynomials and
+        # Alternatives include factoring division polynomials (now implemented) and
         # random sampling (like PARI's ellgroup, but with a milder
-        # termination condition). We should implement these too
+        # termination condition). We should implement the latter too
         # and figure out when to use which.
-        T = self.abelian_group().torsion_subgroup(n)
-        if T.invariants() != (n, n):
-            raise ValueError(f'curve does not have full rational {n}-torsion')
-        return tuple(P.element() for P in T.gens())
+        if algorithm == "abelian_group":
+            T = self.abelian_group().torsion_subgroup(n)
+            if T.invariants() != (n, n):
+                raise ValueError(f'curve does not have full rational {n}-torsion')
+            return tuple(P.element() for P in T.gens())
+        
+        elif algorithm == "division_polynomial":
+            # reduction: n composite -> n prime power -> n prime
+                
+            Ptot,Qtot = self(0), self(0)
+
+            for l,k in n.factor():
+                # find {P,Q} = basis of l-torsion points
+                ltorsion = self(0).division_points(l)
+                if len(ltorsion) < l**2:
+                    raise ValueError(f'curve does not have full rational {l}-torsion')
+                
+                # the first element of 0.division_points(l) is 0 itself, the others have order l
+                P = ltorsion[1]
+                for Q in ltorsion[2:]:
+                    if P.weil_pairing(Q, l) != 1:
+                        break
+                else:
+                    assert False # shouldn't happen
```

Comment:
If you are going to do an if, else like this then have an error a user can undersatnd. If False really can never be reached, the code can be removed.

Reply:
I disagree with removing it: This kind of assertion is great for bug-catching. It *should* never get executed mathematically speaking, so if this point gets reached anyway, we want some kind of alarm going off to indicate that something's wrong.

I don't disagree that adding a human-readable message might be helpful, though: Something like `assert False, "bug: this code should never be reached"` should do the trick.

(Also, in principle, [a compiler could use such statements for optimizations](https://en.cppreference.com/w/cpp/utility/unreachable), but I don't think CPython does this.)

Reply:
I have seen this pattern a few times, and I think I have written code like this too. I wonder if this indicates we should have a `UnreachableError` type? (This should be in base Python, but oh well)

Reply:
Well, we have that: It's called `AssertionError` 😉.